fcrepo-kernel-impl/src/main/java/org/fcrepo/kernel/impl/services/AbstractService.java

```diff
@@ -296,13 +296,51 @@ protected void lockArchivalGroupResourceFromParent(final Transaction tx,
         if (parentId != null && !parentId.isRepositoryRoot()) {
             final var parentHeaders = pSession.getHeaders(parentId, null);
             if (parentHeaders.isArchivalGroup()) {
-                tx.lockResource(parentId);
+                tx.lockResourceAndGhostNodes(parentId);
             } else if (parentHeaders.getArchivalGroupId() != null) {
-                tx.lockResource(parentHeaders.getArchivalGroupId());
+                tx.lockResourceAndGhostNodes(parentHeaders.getArchivalGroupId());
             }
         }
     }
 
+    /**
+     * Check to lock ghost nodes for resourceId depending on the status of the checkId. This allows us to
+     * check a parent for creating new resources, or the actual resource for updating/deleting existing resources.
+     *
+     * @param tx The transaction
+     * @param pSession The persistent storage session
+     * @param resourceId The ID we are locking
+     * @param checkId The ID to check status of
+     */
+    protected void lockResourceDependingOtherStatus(final Transaction tx,
+                                                    final PersistentStorageSession pSession,
+                                                    final FedoraId resourceId,
+                                                    final FedoraId checkId) {
+        final var headers = pSession.getHeaders(checkId, null);
+        if (checkId.isRepositoryRoot() || (!headers.isArchivalGroup() && headers.getArchivalGroupId() == null &&
+                !checkId.isAcl() && !checkId.isDescription())) {
+            // The resource's check ID is the root or is not an AG or part of an AG and is not an ACL or binary
+            // description. So we lock it with ghost node checking
+            tx.lockResourceAndGhostNodes(resourceId);
+        } else {
+            // The checkID is an AG or part of an AG that should have been locked by a call to
+            // lockArchivalGroupResource or lockArchivalGroupResourceFromParent.
+            tx.lockResource(resourceId);
```

Comment:
I think we _do_ want to look for ghosts if it is an AG part. For example, if you have AG `a` and add part `a/b/c` then you'll want to lock the ghost node `a/b`

Reply:
Above we lock ghost nodes if `headers.getArchivalGroupId() != null` (ie. is part of an AG). You did not think that was necessary.

Here we lock ghost nodes if `headers.getArchivalGroupId() == null` (ie. is not part of an AG) and you do think its necessary.

The inversion of `==` to `!=` is confusing me a little so my logic could be incorrect. Also the inclusion of the `lockArchivalGroupResource` and `lockArchivalGroupResourceFromParent` methods cause me some hesitation.

Reply:
Or wait... I think I worked out what you mean